client.go

```diff
@@ -13,244 +13,123 @@ import (
-		return nil
-	}
-}
-
-// MaxIdleConns is configuration option to pass to Client
-// it changes the maximum idle connections that the client will keep
-// in a pool
-func MaxIdleConns(n int) func(*Client) error {
-	return func(c *Client) error {
-		if n < 0 {
-			return ErrInvalidValue
-		}
-		c.maxIdleConns = n
-		return nil
-	}
+	logWriter             io.Writer
+	maxIdleConns          int
+	maxIdleConnsPerHost   int
+	redirectFunc          func(*http.Request, []*http.Request) error
+	responseHeaderTimeout time.Duration
+	tlsHandshakeTimeout   time.Duration
+	transport             *http.Transport
 }
 
-// MaxIdleConnsPerHost is configuration option to pass to Client
-// it changes the maximum idle connections that the client will keep in a pool
-// for a single host.
-// If not specified this will be the same as MaxIdleConns.
-func MaxIdleConnsPerHost(n int) func(*Client) error {
-	return func(c *Client) error {
-		if n < 0 {
-			return ErrInvalidValue
-		}
-		c.maxIdleConnsPerHost = n
-		return nil
-	}
+// Client returns the *http.Client as it was initialized by the constructor
+func (c *client) Client() *http.Client {
+	return c.client
 }
 
-// Logger is configuration option to pass to Client
-// it changes where the debug info is written
-// (ioutil.Discard by default)
-func Logger(w io.Writer) func(*Client) error {
-	return func(c *Client) error {
-		c.logWriter = w
-		return nil
-	}
+// Close is a cleanup function that makes the client close any idle connections
+func (c *client) Close() {
+	c.log.Printf("Close client %#v with all idle connections", c)
+	c.transport.CloseIdleConnections()
 }
 
-// LogPrefix is configuration option to pass to Client
-// to change the prefix used in Clients log output.
-// This can be useful when one is using several httpclients.
-func LogPrefix(p string) func(*Client) error {
-	return func(c *Client) error {
-		c.logPrefix = p
-		return nil
+// New configures and returns a new instance of Client
+func New(options ...Option) (Client, error) {
+	c := newDefaultClient()
+	if err := c.setOptions(options...); err != nil {
+		return nil, err
 	}
+	return c, c.init()
 }
 
-// WithRoundTripper is configuration option to pass to Client
-// this will change the http.RoundTripper that the client will use
-// note use of this renders the use of httpclient pointless
-// since if you are managing your own transports
-// you might as well use net/http directly
-// however this option is useful for unittests
-// e.g. when using httpmock to verify expected traffic
-// when testing some code that uses httpclient
-func WithRoundTripper(t http.RoundTripper) func(*Client) error {
-	return func(c *Client) error {
-		c.customRoundTripper = t
-		return nil
-	}
+// return a new *client with default values
+func newDefaultClient() *client {
+	c := new(client)
+	c.setDefaults()
+	return c
 }
 
-func (c *Client) setDefaults() {
-	c.headers = make(http.Header)
+// set *sensible* default values on the *client
```

Comment:
```suggestion
// set sensible default values on the *client
```

The `*` can be a bit confusing when it's also used to denote a pointer type.